### PR TITLE
Feature/optimize clusters

### PR DESCRIPTION
**Pull request by TobiHSR**

### CODE REVIEW

File: backend/app/services/annoy_service.py [hunk 1]
```diff
@@ -36,30 +42,41 @@ def get_similar_images_for_multiple_ref_images(vectors) -> Dict[str, float]:
...
+        image_duplicate_ids, image_id_to_vector
+    )
+    build_annoy_tree(vectors_without_duplicates)
+    image_id_to_cluster_id: Dict[int, int] = find_neighbor_cluster_for_images(
+        image_id_to_vector, similarity_matrix
     )
-    if len(allocated_images) > 0:
-        save_clusters(overview_ids, allocated_images)
-    print("Cluster build completed")
+    matrix_indexes_for_overview_ids: {int} = find_image_ids_for_overview(
```
Comment (marodev) [inline backend/app/services/annoy_service.py]: {int} does not exists --> Set[int] or List[int] ?

File: backend/app/services/annoy_service.py [hunk 2]
```diff
@@ -76,89 +93,144 @@ def create_similarity_matrix(vectors):
...
+            duplicates.append(duplicated_image + 1)
+            similarity_matrix[duplicated_image] = similarity_matrix[:, duplicated_image]
+    print(f"Similarity matrix builded in {time.time()-start} seconds")
     return similarity_matrix, duplicates
 
 
-def find_image_ids_for_overview(similarity_matrix: ndarray) -> List[int]:
-    overview_distances: List[ndarray] = []
-    overview_ids: List[int] = []
+def valid_image(sm: ndarray, indexes: {int}, image_index: int):
```
Comment (marodev) [inline backend/app/services/annoy_service.py]: {int} does not exists --> Set[int] or List[int] ?

File: backend/app/services/annoy_service.py [hunk 3]
```diff
@@ -76,89 +93,144 @@ def create_similarity_matrix(vectors):
...
+def ignore_cell(sm: ndarray, y: int, x: int):
+    sm[y][x] = 1
+
+
+def ignore_index(sm: ndarray, index):
+    sm[index] = 1
+    sm[:, index] = 1
+
+
+def find_image_ids_for_overview(similarity_matrix: ndarray) -> {int}:
```
Comment (marodev) [inline backend/app/services/annoy_service.py]: same as above

File: backend/app/services/annoy_service.py [hunk 4]
```diff
@@ -76,89 +93,144 @@ def create_similarity_matrix(vectors):
...
+    start = time.time()
+    matrix_indexes_for_overview_ids: {int} = set([])
```
Comment (marodev) [inline backend/app/services/annoy_service.py]: : Set[int] = []

File: backend/app/services/annoy_service.py [hunk 5]
```diff
@@ -76,89 +93,144 @@ def create_similarity_matrix(vectors):
...
-    # max_value: float = similarities.max()
-    for image_index in similar_image_indexes[0]:
-        # max_image_id = where(similarities == max_value)[0][0]
-        # if max_value > config.group_similarity():
-        return get_matches(
-            similarity_matrix, image_index, allocated_vectors, cluster_id
-        )
-    # else:
-    return allocated_vectors
+        min_value_in_matrix: float = similarity_matrix.min()
```
Comment (marodev) [inline backend/app/services/annoy_service.py]: min_value_in_matrix already defined --> min_value_in_matrix = similarity_matrix.min()

File: backend/app/services/annoy_service.py [hunk 6]
```diff
@@ -76,89 +93,144 @@ def create_similarity_matrix(vectors):
...
+    return vector_index_to_image_id[matrix_index]
+
+
+def get_matches(
+    vector_index_to_image_id: Dict[int, int],
+    index_for_matrix: int,
+    similarity_matrix: ndarray,
+    image_id_to_cluster_id: Dict[int, int],
+    cluster_id: int,
+    allocated_images: {int},
```
Comment (marodev) [inline backend/app/services/annoy_service.py]: allocated_images: Set[int]

File: backend/app/services/annoy_service.py [hunk 7]
```diff
@@ -76,89 +93,144 @@ def create_similarity_matrix(vectors):
...
 
 
 def find_neighbor_cluster_for_images(
-    vectors: Dict[int, ndarray], similarity_matrix: ndarray
+    image_id_to_vector: Dict[int, ndarray], similarity_matrix: ndarray
 ) -> Dict[int, int]:
-    allocated_images: Dict[int, int] = {}
+    start = time.time()
+    image_id_to_cluster_id: Dict[int, int] = {}
+    allocated_images: {int} = set([])
```
Comment (marodev) [inline backend/app/services/annoy_service.py]: allocated_images: Set[int] = {}

File: backend/app/services/annoy_service.py (hunk 5 above)
Comment (TobiHSR) [reply backend/app/services/annoy_service.py]: Did not get that?

File: backend/app/services/annoy_service.py (hunk 7 above)
Comment (TobiHSR) [reply backend/app/services/annoy_service.py]: https://stackoverflow.com/questions/6130374/empty-set-literal

File: backend/app/crud/image_manager.py [hunk 8]
```diff
@@ -83,7 +84,9 @@ def get_image_name_by_id(image_id) -> str:
 
 
 @database.connection_context()
-def save_clusters(overview_image_ids: List[int], allocated_images: Dict[int, int]):
+def save_clusters(overview_image_ids: ndarray, allocated_images: Dict[int, int]):
```
Comment (marodev) [inline backend/app/crud/image_manager.py]: def save_clusters(overview_image_ids: ndarray, allocated_images: Dict[int, int]):

def save_clusters(overview_image_ids: ndarray, allocated_images: Dict[int, int]): -> None

**Pull request merged**